### PR TITLE
fix(data): make finetuning batch sampler epoch-aware on checkpoint resume

### LINKED ISSUE #4565: [bug] Abnormal loss fluctuation when resuming training from checkpoint
### Problem

I am performing SFT on a 236B model with DeepSeek V3 architecture. The training was interrupted at around step 25xx. I resumed training from step 2448, which is approximately at the 1.5th epoch. After resuming, the loss dropped to around 0.16 and remained at that level until 200 step before the end of the second epoch，then it raise to 0.2。 Before the interruption, the loss during the second epoch was consistently around 0.2.

I suspected that the data shuffling might be problematic when resuming from checkpoint. To verify this, I tried resuming from step 1632, which is exactly at the end of the first epoch — essentially restarting the second epoch from scratch. After resuming, the loss was around 0.23, which is significantly higher than the normal value of 0.2,  after 100 steps it fall to 0.2。

As shown in the figure below, the red-highlighted region indicates where we resumed training from the checkpoint at step 2448. The loss dropped from slightly below 0.2 to around 0.16. It then returned to approximately 0.2 near the end of the second epoch. At the start of the third epoch, the loss decreased to 0.1 and returned to normal behavior.

Image

### Minimal repro

```shell
this is my script:

def create_finetune_config():
    model = AutoBridge.from_hf_pretrained("hf_model_path").to_megatron_provider(
            load_weights=False
        )
    return ConfigContainer(
        model= model,
        ddp=DistributedDataParallelConfig(
            check_for_nan_in_grad=True,
            grad_reduce_in_fp32=True,
            overlap_grad_reduce=True,
            overlap_param_gather=True,
            bucket_size=100000000,
        ),
        train=TrainingConfig(
            micro_batch_size=1,
            global_batch_size=GLOBAL_BATCH_SIZE,
            train_iters=MAX_STEP,

        ),
        optimizer=OptimizerConfig(
            optimizer="dist_muon",
            lr=2.4e-5,
            min_lr=2.4e-6,
            muon_extra_scale_factor = 0.2,
            weight_decay = 0.1,
            adam_beta1 = 0.9,
            adam_beta2 = 0.95,
            clip_grad = 1.0,
            bf16 = True,
            fp8_recipe = None,
            adam_eps = 1e-8,
            use_distributed_optimizer = False,      # True, ...muon must be false
            overlap_param_gather = True,            # keep same with ddp
        ),
        scheduler=SchedulerConfig(
            start_weight_decay=0.1,
            end_weight_decay=0.1,
            weight_decay_incr_style="constant",
            lr_decay_style="cosine",
            lr_warmup_iters=50,
        ),
        # Finetuning dataset instead of pretraining dataset
        dataset=FinetuningDatasetConfig(
            dataset_root="path_data_root",
            seq_length=SEQ_LENGTH,
            do_validation=False,
            do_test=False,
            # Optional: packed sequence support
            packed_sequence_specs=PackedSequenceSpecs(
                 packed_sequence_size=SEQ_LENGTH,
                packed_train_data_path=PACKED_TRAIN_DATA,
                packed_val_data_path=PACKED_VAL_DATA,
            ),
            dataset_kwargs = {"pad_to_max_length": True}
        ),
        tokenizer=TokenizerConfig(
            tokenizer_type="HuggingFaceTokenizer",
            tokenizer_model="tokenizer path",
        ),
checkpoint=CheckpointConfig(
            pretrained_checkpoint=RESTORE_PATH,
            save=CKPT_DIR,
            load=CKPT_DIR,
            save_interval=EVERY_N_TRAIN_STEPS,
            most_recent_k=100,
            async_save=True,
            fully_parallel_save=True,
            fully_parallel_load=True,
            dist_ckpt_strictness="log_unexpected",
        ),
        mixed_precision="bf16_mixed",
        rng=RNGConfig(seed=1234),
        comm_overlap=CommOverlapConfig(
            tp_comm_overlap=False,
        ),
        logger=LoggerConfig(
            log_interval=1,
            tensorboard_dir=CKPT_DIR,
        ),
    )
if __name__ == "__main__":
    cfg = create_finetune_config()
   
    cfg.model.fp8_recipe = None
    cfg.model.params_dtype = torch.float32
    cfg.optimizer.params_dtype = torch.float32
    cfg.optimizer.fp8_recipe = None

    # attention
    cfg.model.num_attention_heads = 128
    cfg.model.kv_channels = 40
    cfg.model.num_query_groups = 128
    cfg.model.num_layers = 60
    cfg.model.hidden_size = 5120
    cfg.model.ffn_hidden_size = 12288
    cfg.model.vocab_size = 152064

    #mtp
    cfg.model.mtp_num_layers = None
    cfg.model.mtp_loss_scaling_factor  = 0.1

    #moe
    cfg.model.num_moe_experts = 160
    cfg.model.moe_ffn_hidden_size = 1536
    cfg.model.moe_shared_expert_intermediate_size = 3072
    cfg.model.moe_layer_freq= [0] * 1 + [1] * 59
    cfg.model.moe_token_dispatcher_type = "flex"
    cfg.model.moe_enable_deepep = True
    cfg.model.moe_shared_expert_overlap = False           # moe_shared_expert_overlap only works with alltoall token dispatcher
    cfg.model.moe_aux_loss_coeff = 1e-4
    cfg.model.moe_z_loss_coeff = None
    cfg.model.moe_router_topk = 6
    cfg.model.moe_router_num_groups = 1
    cfg.model.moe_router_group_topk = 1
    cfg.model.moe_router_topk_scaling_factor = 2.827
    cfg.model.make_vocab_size_divisible_by = 512
    cfg.model.moe_router_score_function = "sigmoid"
    cfg.model.moe_router_enable_expert_bias = True
    cfg.model.moe_router_bias_update_rate = 1e-3
    cfg.model.moe_router_pre_softmax = True
    cfg.model.moe_grouped_gemm = True
    cfg.model.moe_router_load_balancing_type = "seq_aux_loss"
    cfg.model.moe_router_dtype = "fp32"

    #MLA
    cfg.model.multi_latent_attention=True
    cfg.model.kv_lora_rank = 512
    cfg.model.q_lora_rank = 1536
    cfg.model.qk_head_dim = 128
    cfg.model.qk_pos_emb_head_dim = 64
    cfg.model.v_head_dim = 128
    cfg.model.mscale = 1.0
    cfg.model.mscale_all_dim = 1.0

    # rope
    cfg.model.max_position_embeddings = SEQ_LENGTH
    cfg.model.seq_length = SEQ_LENGTH
    cfg.model.rotary_base = 5000000
    cfg.model.rope_type="rope"

    # recompute
    cfg.model.recompute_granularity = 'full'
    cfg.model.recompute_method = 'uniform'
    cfg.model.recompute_num_layers = 1

    # 3d parallel
    cfg.model.sequence_parallel=True
    #cfg.model.num_layers_in_first_pipeline_stage=8
    cfg.model.num_layers_in_last_pipeline_stage=4
    cfg.model.tensor_model_parallel_size=16
    cfg.model.pipeline_model_parallel_size=8
    cfg.model.expert_model_parallel_size=8
    cfg.model.expert_tensor_parallel_size=1
    cfg.model.virtual_pipeline_model_parallel_size=4
    cfg.model.context_parallel_size=2
    cfg.model.batch_p2p_comm=True

    # fusion
    cfg.model.attention_softmax_in_fp32 = True
    cfg.model.moe_permute_fusion = False
    cfg.model.apply_rope_fusion = False
    cfg.model.bias_activation_fusion = False
    cfg.model.bias_dropout_fusion = False
    cfg.model.masked_softmax_fusion = False
    cfg.model.gradient_accumulation_fusion = False
    cfg.model.cross_entropy_loss_fusion = False
    cfg.model.persist_layer_norm = False
    #cfg.model.params_dtype = torch.float32
    cfg.model.init_method_std = 0.006
    cfg.model.embedding_init_method_std = 0.006


    # when cp > 1，calculate_per_token_loss must be True
    cfg.model.calculate_per_token_loss = True if cfg.model.context_parallel_size > 1 else False
    cfg.model.sequence_parallel = True if cfg.model.tensor_model_parallel_size > 1 else False
    cfg.dataset.packed_sequence_specs.pad_seq_to_mult = cfg.model.context_parallel_size * 2 if  cfg.model.context_parallel_size > 1 else 1

    print_rank_0("-----all transformer config-------")
    print_rank_0(cfg.model)

    finetune(cfg, forward_step_func=forward_step)
```

### Expected behavior

The loss is expected to remain stable when resuming from a checkpoin

### Affected area

area:training

### Regression?

Yes

### Environment

_No response_

### Logs

```shell

```

### PR DESCRIPTION
**Pull request by Achyuthan-S**
# What does this PR do ?

Fixes abnormal loss when resuming fine-tuning from a checkpoint (#4565). The
fine-tuning batch sampler froze `consumed_samples` and re-applied the resume
offset on every epoch, so after a mid-epoch resume each later epoch skipped the
start of the dataset and replayed only the tail — repeated data that depressed
the loss. This makes `MegatronPretrainingBatchSampler` epoch-aware (and adds
deterministic per-epoch reshuffling), mirroring `MegatronPretrainingRandomSampler`.

# Changelog

- `data/samplers.py`: `MegatronPretrainingBatchSampler.__iter__` now advances
  `consumed_samples` as it yields and computes the within-epoch offset against the
  whole-global-batch multiple, so each `cyclic_iter` re-iteration starts a fresh
  full epoch instead of permanently re-applying the resume offset and replaying
  the tail.
- `data/samplers.py`: add `shuffle`/`seed` params for deterministic, seed- and
  epoch-derived per-epoch reshuffling (resume-reproducible). Threaded through
  `build_pretraining_data_loader`. Both default to off → backward compatible.
- `data/loaders.py`: enable shuffle for the fine-tuning (`batch`) **train**
  dataloader, seeded by the dataset seed. Eval/test and pretraining paths are
  unchanged.
- `tests/.../data/test_samplers.py`: regression tests for resume-serves-full-epoch,
  epoch-boundary resume, per-epoch reshuffle, seed determinism, and shuffle resume
  parity. Existing batch-sampler assertions are unchanged.

**Behavior change:** fine-tuning data order now reshuffles each epoch
(deterministic and resume-reproducible). SFT loss curves will shift versus the
previous sequential order. New params are optional and default off.

# GitHub Actions CI

See the [CI section](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md#-running-github-ci) in the Contributing doc for how to trigger the CI. A Nvidia developer will need to approve and trigger the CI for external contributors.

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests? (sampler resume-parity + reshuffle regression tests)
- [x] Did you add or update any necessary documentation? (evaluated — no existing doc covers fine-tuning sampler/resume order; the behavior change is described above. Happy to add a docs note if maintainers prefer.)
- [x] Does the PR affect components that are optional to install? No
  - [x] Reviewer: Does the PR have correct import guards for all optional libraries? N/A

# Additional Information

- Closes #4565
- Suggested labels: `bug`, `area:data`, `needs-more-tests` (this changes SFT data ordering).

### CODE REVIEW

**Comment by yaoyu-33** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/4601/changes/dd99eaedaf1d957931b023555892aceadf754321

**Comment by yaoyu-33** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/4601/commits/4c996b0af3fedb8d32ee597e46811768f09585d0

**Comment by Achyuthan-S** [comment]
Hello @yaoyu-33  , it was great contributing. Thank you for the review.

If you have any other issues that i can contribute to and solve .., it would be great if you tag me to it whenever .

Thanks again !

**Review by liding-nv — APPROVED** [review]
